### PR TITLE
add simple metrics (PSNR, NRMSE)

### PR DESCRIPTION
**Pull request by grlee77**
This PR adds simple metrics:  `mse`, normalized root mean square error (`nrmse`) and peak signal to noise ratio (`psnr`).

These are all pretty trivial to implement, but are still useful to have on hand.  PSNR in particular is widely used as a comparison metric in image denoising/compression papers.  

I realize that the existing `skimage.measure.structural_similarity` (SSIM) metric is generally superior to these in terms of making image comparisons, but I would argue these are still useful to have around, if only for comparing to results from the literature.

### CODE REVIEW

File: skimage/measure/_simple_metrics.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    if not im_true.shape == im.shape:
+        raise ValueError('Input images must have the same dimensions.')
+
+    norm_type = norm_type.lower()
+    if norm_type == 'euclidean':
+        denom = np.sqrt((im_true*im_true).mean())
+    elif norm_type == 'min-max':
+        denom = im_true.max() - im_true.min()
+    elif norm_type == 'mean':
+        denom = im_true.mean()
```
Comment (ahojnnes) [inline skimage/measure/_simple_metrics.py]: Add an "else" branch, raising a `ValueError`. Also add test using `assert_raises`. Thanks.

File: skimage/measure/_simple_metrics.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
+from __future__ import division
+
+import numpy as np
+from ..util.dtype import dtype_range
+
+__all__ = ['mse', 'nrmse', 'psnr']
+
+
+def mse(X, Y):
+    """ compute mean-squared error between two images.
```
Comment (ahojnnes) [inline skimage/measure/_simple_metrics.py]: "compute" -> "Compute", remove leasing space in the line.

File: skimage/measure/_simple_metrics.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    if not X.shape == Y.shape:
+        raise ValueError('Input images must have the same dimensions.')
+    if not X.dtype == Y.dtype:
+        raise ValueError('Input images must have the same dtype.')
+    return np.square(X - Y).mean()
+
+
+def nrmse(im_true, im, norm_type='Euclidean'):
+    """ compute the normalized root mean-squared error between two images.
```
Comment (ahojnnes) [inline skimage/measure/_simple_metrics.py]: "compute" -> "Compute", remove leasing space in the line.

File: skimage/measure/_simple_metrics.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    Parameters
+    ----------
+    im_true : ndarray
+        Ground-truth image.
+    im : ndarray
```
Comment (ahojnnes) [inline skimage/measure/_simple_metrics.py]: Maybe call `im_test`?

File: skimage/measure/_simple_metrics.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    -------
+    nrmse : float
+        The NRMSE metric.
+
+    Notes
+    -----
+    There is no standard method of normalization across the literature [1]_.
+    The methods available here are as follows:
+
+    - 'Euclidean' : normalize by the Euclidean norm of ``im_true``.
```
Comment (ahojnnes) [inline skimage/measure/_simple_metrics.py]: Maybe this fits better into the `norm_type` section?

File: skimage/measure/tests/test_simple_metrics.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    x = np.ones(4)
+    y = np.asarray([0., 2., 2., 2.])
+    assert_equal(nrmse(y, x, 'mean'), 1/np.mean(y))
+    assert_equal(nrmse(y, x, 'Euclidean'), 1/np.sqrt(3))
+    assert_equal(nrmse(y, x, 'min-max'), 1/(y.max()-y.min()))
+
+    assert_equal(nrmse(x, x), 0)
+
+    assert_raises(ValueError, nrmse, x.astype(np.uint8), y)
+    assert_raises(ValueError, nrmse, x[:-1], y)
```
Comment (ahojnnes) [inline skimage/measure/tests/test_simple_metrics.py]: Please, add `if __name__ ==...` section as in other unit test files.

File: skimage/measure/_simple_metrics.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
+from __future__ import division
```
Comment (ahojnnes) [inline skimage/measure/_simple_metrics.py]: No reason to use the leading underscore for the file name.

**Comment by ahojnnes** [comment]
@grlee77 This is definitely useful, thanks. I left you some comments.

**Comment by grlee77** [comment]
For `psnr`, would it be preferable default to ``dynamic_range = 1`` for float32 or float64 inputs?  I would guess that is the more common use case, but the current implementation uses `dtype_range` which will be [-1, 1] ``dynamic_range = 2``.

**Comment by ahojnnes** [comment]
@grlee77 Maybe we should not set a default value at all, thereby forcing the user to think about the correct value?

**Comment by jni** [comment]
I'm very much in favour of a range of 1. 99% of the use cases here will be images that have been read in from a uint8 and converted using img_as_float, and so will have range [0, 1].

**Comment by grlee77** [comment]
For `psnr`, I changed the range to be the positive portion of the range returned by `dtype_range` and issue a warning if values exceeding this range are found in `im_truth`.  Does that sound reasonable?  As @jni mentioned, this default should cover 99% of the use cases

Also,  `nrmse` and `mse` now also have a proper conversion to float before performing the calculations.

**Comment by jni** [comment]
@grlee77 What I would do is make it 1 for float images and 2 for float images containing any negative values. But this is fine too.

The failing test is pretty funny:
```
Arrays are not almost equal to 7 decimals
 ACTUAL: 22.449691648492035
 DESIRED: 22.44969236505311
```

I presume this difference is not important, in which case, you can change the precision of the check.

**Comment by grlee77** [comment]
I reduced the precision of the check to 5 decimal places.  Usually only 1-2 decimal places are used in reporting PSNR values so that should be plenty.

The failure was due to a change in how the conversion to floating point is done.  Originally everything was converted to float64, but after f4773e303de43f8c5f1193fd23bb841394c77134 `uint8` inputs are converted to float32 via `np.result_type`, so the precision no longer matches the float64 input case.

**Comment by grlee77** [comment]
Python 2.7 with older `numpy` 1.6.1 still fails some of the precision checks.  Somehow the agreement is only to 1 decimal place there, but is passing the same tests at 4-5 decimal places on the other versions!  

I can try just changing back to float64 and see if that fixes it, but any ideas why this would be?

**Comment by jni** [comment]
@grlee77 It could well be that numpy updated their `mean` function to use a more numerically-stable summation algorithm after 1.6. There's talk of bumping the requirement to numpy 1.7 before scikit-image 0.12, so you might be off the hook after a rebase.

Other than that I have the following style change recommendations, in decreasing order of importance:

- X, Y -> im_test, im_true / ar_test, ar_true
- foo.mean() -> np.mean(foo) (ditto for min, max, etc)
- mse -> mean_squared_error. I think PSNR is much more commonly used as an acronym, whereas MSE is often referred to by its full name, e.g. in conversation. To me, just "mse" is hard to parse, especially because of the lowercase.

@stefanv can chime in on these.

**Comment by stefanv** [comment]
+1 from me
​

**Comment by grlee77** [comment]
@jni
I agree with the suggestions, but two quick questions:
- can NRMSE stay as `nrmse`?  The full expansion of that acronym starts to become inconveniently long to type.
- How about `im1`, `im2` in place of `X`, `Y` instead since it does not matter what order the images are input to `mse` or the utility functions that were using (X, Y)?  (The X, Y notation was copied from the previously existing `structural_similarity` metric)

**Comment by grlee77** [comment]
according to numpy/numpy#5681, the precision of mean was improved in numpy 1.9.  However, even for 1.6, one can set `dtype=np.float64` to use 64-bit precision internally for the accumulation.  There is even an example in the docs:  http://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.mean.html
so I guess I could go back to the `np.result_type`-based conversion of the images and use `dtype=np.float64` for the calls to `np.mean`

**Comment by JDWarner** [comment]
@grlee77 sounds like a great workaround for NumPy 1.6. 

For this subset of functions intended to compare two images, perhaps we should move toward an API based on a prefix like `skimage.measure.compare_*` with the wildcard being a brief name. For instance, at present:

* `compare_similarity`
* `compare_mse`
* `compare_nrmse`
* `compare_psnr`

This would involve deprecating `structural_similarity` but might be worth it for accessibility.

**Comment by stefanv** [comment]
@JDWarner Very good observation, thanks.

**Comment by jni** [comment]
I love the mean precision workaround. Thanks for that research @grlee77!

I'm -1 on `compare`, as it is proposed at least, because the name is misleading. To me, `compare_mse` means I will compare two different MSE values or a list of MSE values. `compare_with` or `compare_by` would be a more acceptable prefix, but the function names then become cumbersome.

I think the namespace itself does an ok job, e.g. `measure.mean_squared_error(im1, im2)`. That reads quite a bit more nicely than `measure.compare_mse(im1, im2)`. I think APIs should strive to read as closely to plain English as possible. This is a big part of what makes Python so easy to learn.

@grlee77 as to `nrmse`, I would perhaps call it `normalized_root_mse`.

**Comment by stefanv** [comment]
@jni There is some value in grouping these things, I think.  Perhaps a more suitable prefix, such as `difference`?

**Comment by jni** [comment]
@stefanv Sure, but maybe raise a separate issue? I think this discussion doesn't belong in this PR. An entirely new namespace might be warranted, though. `measure`: measure things within an image. `difference`/`compare`: compare different images.

**Comment by stefanv** [comment]
@jni Given that this PR adds three new functions, now is not a bad time to discuss it.  Are you opposed to ``difference`` or ``diff`` as a prefix?

**Comment by grlee77** [comment]
I would suggest `metrics` as a potential namespace as these can all be considered image quality metrics.  I would prefer that to prefixing `compare_` or  `diff_` to each function name.

**Comment by stefanv** [comment]
Either way works for me.

**Comment by grlee77** [comment]
I made the stylistic changes suggested by @jni and am now seeing if the tests pass using `np.mean` with the `dtype=np.float64` argument.

If a consensus develops to move into a different namespace, let me know whether you prefer to do so here or in a separate PR.

**Comment by jni** [comment]
@stefanv because it also involves a deprecation of an existing function. I feel quite strongly that PRs with new features should not be forced to fix existing mistakes.

**Comment by stefanv** [comment]
@jni No, not forcing--but we're always trying to move the design in the right direction.  I don't care if it happens on this PR or another, as long as they both get merged either before or after the release.

**Comment by jni** [comment]
@stefanv Well afaic this one is ready (thanks @grlee77!), but let's wait before merging until either a consensus is reached in #1914 or 0.12 is released.

**Comment by JDWarner** [comment]
Merging this, thanks for your contributions and sticking with us @grlee77! 

I'll address the naming conventions separately tomorrow in a small PR.
